packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts

```diff
@@ -0,0 +1,117 @@
+    SwapQuote,
+    SwapQuoteConsumer,
+    SwapQuoteConsumerOpts,
+} from '../types';
+import { assert } from '../utils/assert';
+import { assetDataUtils } from '../utils/asset_data_utils';
+import { swapQuoteConsumerUtils } from '../utils/swap_quote_consumer_utils';
+
+import { ExchangeSwapQuoteConsumer } from './exchange_swap_quote_consumer';
+import { ForwarderSwapQuoteConsumer } from './forwarder_swap_quote_consumer';
+
+type SmartContractParams =
+    | ExchangeMarketBuySmartContractParams
+    | ExchangeMarketSellSmartContractParams
+    | ForwarderMarketBuySmartContractParams
+    | ForwarderMarketSellSmartContractParams;
+
+type ValidSwapQuoteConsumer = ExchangeSwapQuoteConsumer | ForwarderSwapQuoteConsumer;
+
+export class DynamicSwapQuoteConsumer implements SwapQuoteConsumer<SmartContractParams> {
+    public readonly provider: ZeroExProvider;
+    public readonly networkId: number;
+
+    private readonly _contractWrappers: ContractWrappers;
+    private readonly _exchangeConsumer: ExchangeSwapQuoteConsumer;
+    private readonly _forwarderConsumer: ForwarderSwapQuoteConsumer;
+
+    constructor(supportedProvider: SupportedProvider, options: Partial<SwapQuoteConsumerOpts> = {}) {
+        const { networkId } = _.merge({}, constants.DEFAULT_SWAP_QUOTER_OPTS, options);
+        assert.isNumber('networkId', networkId);
+
+        const provider = providerUtils.standardizeOrThrow(supportedProvider);
+        this.provider = provider;
+        this.networkId = networkId;
+        this._contractWrappers = new ContractWrappers(this.provider, {
+            networkId,
+        });
+
+        this._exchangeConsumer = new ExchangeSwapQuoteConsumer(supportedProvider, options);
+        this._forwarderConsumer = new ForwarderSwapQuoteConsumer(supportedProvider, options);
+    }
+
+    public async getCalldataOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteGetOutputOpts>,
+    ): Promise<CalldataInfo> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.getCalldataOrThrowAsync(quote, opts);
+    }
+
+    public async getSmartContractParamsOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteGetOutputOpts>,
+    ): Promise<SmartContractParamsInfo<SmartContractParams>> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.getSmartContractParamsOrThrowAsync(quote, opts);
+    }
+
+    public async executeSwapQuoteOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteExecutionOpts>,
+    ): Promise<string> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.executeSwapQuoteOrThrowAsync(quote, opts);
+    }
+
+    private async _getConsumerForSwapQuoteAsync(
```

Comment:
Is it tenable to write a test for this function?

Reply:
making the function public; should make it testable